R/literal_coercion_linter.R

```diff
@@ -84,13 +82,31 @@ literal_coercion_linter <- function() {
 
     bad_expr <- xml2::xml_find_all(xml, xpath)
 
+    coercer <- xp_call_name(bad_expr)
+    # tiptoe around the fact that we don't require {rlang}
+    if (any(coercer %in% rlang_coercers) && !requireNamespace("rlang", quietly = TRUE)) {
+      # NB: we _could_ do some extreme customization where each lint
+      #   gets a message according to whether the coercer is from rlang,
+      #   but this seems like overkill. Just use a generic message and move on.
+      lint_message <- paste(
+        "Use literals directly where possible, instead of coercion.",
+        "c.f. 1L instead of as.integer(1) or rlang::int(1), or NA_real_ instead of as.numeric(NA).",
+        "NB: this message can be improved to show a specific replacement if 'rlang' is installed."
+      )
+    } else {
+      coercion_str <- xml2::xml_text(bad_expr)
+      # the linter logic & rlang requirement should ensure that it's safe to run eval() here
+      literal_equivalent_str <- vapply(str2expression(coercion_str), function(expr) deparse1(eval(expr)), character(1L))
```

Comment:
NB: this will return `1` instead of `1.0` for `as.numeric(1)`.
```r
deparse1(eval(quote(as.numeric(1))))
#> [1] "1"
```
Feel free to make this a follow-up though.

Reply:
Hmm using `1.0` assumes `implicit_integer_linter()` is active though right? not sure what our default message should be in that case...

Reply:
Well, I think the general advice to use `1.0` for `as.numeric(1)` seems very good, even if implict integers aren't linted. The generic method seems fine as-is.

Reply:
Added a TODO for now

Reply:
> It could be more geared towards rlang seeing that the base functions will never cause it, though.

Yes and no... the generic will be thrown twice here, for example (if `rlang` is missing):

```r
c(
  as.integer(1),
  rlang::int(1)
)
```

Reply:
Oh, I didn't notice that. Is there harm in vectorizing the message code?

Reply:
Just maintenance overhead. Started writing it like that but code was already getting ugly / requiring careful refactor to handle the two branches without too much duplication.

Stopped because it doesn't seem that likely to occur in practice IMO. Seems rare that user (1) uses `rlang`, but doesn't have it installed in the linting environment and (2) uses `rlang` inconsistently, i.e., `chr()` in one place but `as.integer()` in another

Reply:
Okay, seems fair.